In [12]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
x_train = pd.read_csv('./x_train.csv', index_col='ID')
y_train = pd.read_csv('./y_train.csv', index_col='ID')
test = pd.read_csv('x_test.csv', index_col='ID')
#concaténation des deux dataframes pour le préprocessing sur la colonne ID
data = pd.concat([x_train, test], axis=0)
data.head()

DATE  STOCK  INDUSTRY  INDUSTRY_GROUP  SECTOR  SUB_INDUSTRY     RET_1  \
ID                                                                          
0      0      2        18               5       3            44 -0.015748   
1      0      3        43              15       6           104  0.003984   
2      0      4        57              20       8           142  0.000440   
3      0      8         1               1       1             2  0.031298   
4      0     14        36              12       5            92  0.027273   

    VOLUME_1     RET_2  VOLUME_2  ...    RET_16  VOLUME_16    RET_17  \
ID                                ...                                  
0   0.147931 -0.015504  0.179183  ...  0.059459   0.630899  0.003254   
1        NaN -0.090580       NaN  ...  0.015413        NaN  0.003774   
2  -0.096282 -0.058896  0.084771  ...  0.008964  -0.010336 -0.017612   
3  -0.429540  0.007756 -0.089919  ... -0.031769   0.012105  0.033824   
4  -0.847155 -0.039302 -0.943033  ... -0.038461  -0.277083 -0.012659   

    VOLUME_17    RET_18  VOLUME_18    RET_19  VOLUME_19    RET_20  VOLUME_20  
ID                                                                            
0   -0.379412  0.008752  -0.110597 -0.012959   0.174521 -0.002155  -0.000937  
1         NaN -0.018518        NaN -0.028777        NaN -0.034722        NaN  
2   -0.354333 -0.006562  -0.519391 -0.012101  -0.356157 -0.006867  -0.308868  
3   -0.290178 -0.001468  -0.663834 -0.013520  -0.562126 -0.036745  -0.631458  
4    0.139086  0.004237  -0.017547  0.004256   0.579510 -0.040817   0.802806  

[5 rows x 46 columns]

In [7]:
# 1.1 Valeurs manquantes
print("Valeurs manquantes par colonne :\n", data.isnull().sum())

# 1.2 Statistiques descriptives globales
print("\nStatistiques descriptives (numériques) :\n", data.describe())


Valeurs manquantes par colonne :
 DATE                   0
STOCK                  0
INDUSTRY               0
INDUSTRY_GROUP         0
SECTOR                 0
SUB_INDUSTRY           0
RET_1               2367
VOLUME_1           92215
RET_2               2482
VOLUME_2           91780
RET_3               2529
VOLUME_3           95190
RET_4               2577
VOLUME_4          100196
RET_5               2626
VOLUME_5          105788
RET_6               2642
VOLUME_6          105816
RET_7               2631
VOLUME_7          104962
RET_8               2667
VOLUME_8          105018
RET_9               2722
VOLUME_9          104426
RET_10              2752
VOLUME_10         102558
RET_11              4014
VOLUME_11         100310
RET_12              3352
VOLUME_12          87883
RET_13              3634
VOLUME_13          84494
RET_14              6075
VOLUME_14          89542
RET_15              7009
VOLUME_15          95307
RET_16              6605
VOLUME_16          93613
RET_17          

In [10]:
# 1.3 Nombre d’observations attendu vs réel
print(f"\nNombre d’observations x_train:{x_train.shape[0]}")
print(f"Nombre d’observations x_test :{test.shape[0]}")

# 1.4 Types de colonnes
print("\nTypes de données par colonne :\n", data.dtypes)


Nombre d’observations x_train:418595
Nombre d’observations x_test :198429

Types de données par colonne :
 DATE                int64
STOCK               int64
INDUSTRY            int64
INDUSTRY_GROUP      int64
SECTOR              int64
SUB_INDUSTRY        int64
RET_1             float64
VOLUME_1          float64
RET_2             float64
VOLUME_2          float64
RET_3             float64
VOLUME_3          float64
RET_4             float64
VOLUME_4          float64
RET_5             float64
VOLUME_5          float64
RET_6             float64
VOLUME_6          float64
RET_7             float64
VOLUME_7          float64
RET_8             float64
VOLUME_8          float64
RET_9             float64
VOLUME_9          float64
RET_10            float64
VOLUME_10         float64
RET_11            float64
VOLUME_11         float64
RET_12            float64
VOLUME_12         float64
RET_13            float64
VOLUME_13         float64
RET_14            float64
VOLUME_14         float64
RET_15  

In [14]:
# 2.1 Distribution de la cible
ret_counts = y_train['RET'].value_counts().sort_index()
print("\nRépartition des classes RET :\n", ret_counts)


Répartition des classes RET :
 RET
False    209749
True     208846
Name: count, dtype: int64


In [15]:
# 2.2 Statistiques descriptives RET_1…RET_20 et VOLUME_1…VOLUME_20
cols_ret    = [f"RET_{i}"    for i in range(1,21)]
cols_volume = [f"VOLUME_{i}" for i in range(1,21)]

print("\nStatistiques RET_1–RET_20 :\n", x_train[cols_ret].describe().T[['mean','std','min','max']])
print("\nStatistiques VOLUME_1–VOLUME_20 :\n", x_train[cols_volume].describe().T[['mean','std','min','max']])

# 2.3 Cardinalité des variables catégorielles
cat_cols = ['STOCK','INDUSTRY','INDUSTRY_GROUP','SUB_INDUSTRY','SECTOR']
for c in cat_cols:
    print(f"{c} : {x_train[c].nunique()} catégories")


Statistiques RET_1–RET_20 :
             mean       std       min         max
RET_1   0.001383  0.031311 -0.845324    1.444990
RET_2   0.000973  0.030987 -0.770751    1.427746
RET_3   0.002153  0.031332 -0.740406    3.086617
RET_4  -0.000679  0.031224 -0.863554    2.243749
RET_5   0.000358  0.031886 -0.792839    1.491705
RET_6  -0.000261  0.031311 -0.723077    2.810885
RET_7   0.000330  0.030966 -0.660297    1.512062
RET_8   0.000124  0.031707 -0.806159    2.290456
RET_9  -0.000621  0.032899 -0.896307    2.471602
RET_10  0.000005  0.031689 -0.753419    1.999950
RET_11 -0.000781  0.030733 -0.755627    1.355075
RET_12  0.000914  0.187321 -0.807084  118.999563
RET_13  0.001349  0.032356 -0.828125    1.722330
RET_14  0.002468  0.030380 -0.827586    1.285713
RET_15  0.001846  0.032575 -0.816000    1.619833
RET_16  0.001074  0.029538 -0.765478    1.028902
RET_17  0.001642  0.031914 -0.825014    1.707736
RET_18  0.001131  0.031435 -0.876157    6.000004
RET_19 -0.000798  0.030738 -0.880000   

## First part

In [22]:
# 1. Calcul moyenne RET_1 groupée par SECTOR, DATE
mean_by_sector_date = (
    train
    .reset_index()  # Ramène 'ID' comme colonne si ce n’est pas déjà le cas
    .groupby(['SECTOR', 'DATE'])['RET_1']
    .mean()
    .reset_index()  # Nous donne SECTOR, DATE, MEAN_RET1
    .sort_values(['SECTOR', 'DATE'])
)

# 2. Ajout de la statistique décalée (shift=1) par SECTOR
mean_by_sector_date['MEAN_RET1_SECTOR_DATE'] = (
    mean_by_sector_date
    .groupby('SECTOR')['RET_1']
    .shift(1)
)

# 3. Merge avec le DataFrame d’origine
train_fe = (
    train
    .reset_index()  # Si l’index n’est pas déjà une colonne nommée 'ID'
    .merge(mean_by_sector_date[['SECTOR', 'DATE', 'MEAN_RET1_SECTOR_DATE']],
           on=['SECTOR', 'DATE'],
           how='left')
    .set_index('ID')  # Recrée l’index d’origine si besoin
)

print(train_fe[['SECTOR', 'DATE', 'RET_1', 'MEAN_RET1_SECTOR_DATE']].head())


    SECTOR  DATE     RET_1  MEAN_RET1_SECTOR_DATE
ID                                               
0        3     0 -0.015748                    NaN
1        6     0  0.003984                    NaN
2        8     0  0.000440                    NaN
3        1     0  0.031298                    NaN
4        5     0  0.027273                    NaN


In [ ]:
target = 'RET'

n_shifts = 10  # If you don't want all the shifts to reduce noise
features = ['RET_%d' % (i + 1) for i in range(n_shifts)]
features += ['VOLUME_%d' % (i + 1) for i in range(n_shifts)]
features += new_features  # The conditional features
train[features].head()

We have a ton of rows—418,595 for training and 198,429 for testing. Each row is a stock, a day, with 46 descriptive things about it.

Quick Summary:
No missing data (no NaNs). Phew.
Columns are clean: we have dates, stock/sector names (that's the categorical stuff), and numbers (the 20 days of return RET and volume VOLUME).
The variable to predict (RET, the gain/loss sign) is perfectly balanced: like 50% gain, 50% loss. 
The returns (RET): super calm, mean is near zero, not much movement (standard deviation around 0.02).
The volumes (VOLUME): they go crazy sometimes, with really huge values (right-skewed distribution).
There are 500 stocks and 5 big sectors to deal with.
~20 INDUSTRY, ~10 INDUSTRY_GROUP, ~50 SUB_INDUSTRY, ~5 SECTOR.

We combined the training and test inputs. This helps us apply uniform pre-processing.
For categorical features, we used One-Hot Encoding on industries and sectors. We used simple integer encoding for the stock IDs.
We standardized all numerical features. This means we gave them zero mean and unit variance. This makes them comparable.
We did not need to handle initial NaNs. However, new features we created later did introduce some NaNs. We imputed these using the global mean.
We generated new features. These are conditional aggregated features. For example, we calculated the mean of returns by SECTOR. This helps capture sector trends and reduce noise.

This is a binary classification task. We need to predict the sign of the residual return.
We chose Logistic Regression as our baseline. It uses L2 regularization.
This model is easy to interpret. It's also fast and efficient for high-dimensional data. It sets a simple reference point.